### PR TITLE
Fix static lib registration

### PR DESCRIPTION
**Pull request by randombit**
The ongoing support issues around the static library not behaving as expected are trouble; anyone who builds a static lib with any kind of custom build system will end up with a misbehaving library.

The issues with the registration system come from the static linker not considering the objects as needed at link time as there is no explicit reference in the code, so the objects are discarded and then lookups fail.

Instead group the registrations by type and in the same file include something which the linker will know is required, for example the destructor of the base type. Thus simply referencing a base type is enough to ensure the object registrations will occur regardless of static/shared.

With shared libs there is no real change here beyond a lot of extra #ifdefs since all objects are linked into the .so and are always visible and included. With static libs, more objects are pulled in than are strictly necessary, but much fewer than with the amalgamation.

Tests now pass with static non-amalgamation builds (verified by circleci change), so remove the restriction set in configure.py that requires amalgamation builds for static libs.

### CODE REVIEW

**Comment by cdesjardins** [comment]
Nice work, I will certainly be giving this a try within the week! I still wonder if this issue could be resolved in a cleaner way with an API change. I haven't fully worked through it, so it could be a waste of time, but it might be a fun experiment one day.

**Comment by randombit** [comment]
Great. I'm going to leave this one open for review and testing for a while I think.

**Comment by cdesjardins** [comment]
I have started playing with this and it seems better, `get_block_cipher()` appears to work, and `get_hash_function()` appears to work, of course my testing is limited so far because there is still an issue with RNG:
```
#include "botan/botan.h"

int main(int argc, char** argv)
{
    std::unique_ptr<Botan::RandomNumberGenerator> rng(new Botan::Serialized_RNG());
    return 0;
}
```

When I run the above program I get:
```
terminate called after throwing an instance of 'Botan::Algorithm_Not_Found'
  what():  Could not find any algorithm named "HMAC(SHA-512)"
Aborted
```
I am using:
```
$ g++ --version
g++ (Debian 4.9.2-10) 4.9.2
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

and configure with:
```
./configure.py --disable-shared --disable-modules=tls --prefix=$PWD/../install --libdir=$PWD/../install/lib/botan/Debug --build-mode=debug --disable-avx2 --maintainer-mode
```

I still haven't tried it on Windows yet, and I am totally open to suggestion if there is a preferred/better/working way to get a rng. 

p.s. all tests in botan-test pass for me with the above configuration, and I do have the correct branch checked out, the git sha1 is: `d83ef010522373a6f8ed3876c812b18b55513103`

**Comment by randombit** [comment]
There were some corner cases where the registration objects would not be pulled in. That should be fixed with my most recent push, and your test now works for me.

**Comment by cdesjardins** [comment]
Alright, I had some more time to try this out again, and I tested with the e2e0f8f checkin, I haven't tried the "Initial header cleanups" patch because it wasn't there when I updated my branch this morning, and now it looks like it has CI failures anyways... 

All that aside, I can confirm that I am able to connect to ssh servers with e2e0f8f, and the same configure as above using a bunch of different algorithms, and also using encrypted private keys. So for me this is a win, obviously it isn't a full test of everything in botan, but the subset that I use seems to totally work now with a static non-amalgam build.

Thanks!

File: src/lib/pbkdf/pbkdf.h [hunk 1]
```diff
@@ -147,6 +161,20 @@ class BOTAN_DLL PBKDF
...
+/**
+* Password based key derivation function factory method
+* @param algo_spec the name of the desired PBKDF algorithm
+* @return pointer to newly allocated object of that type
+*/
+inline PBKDF* get_pbkdf(const std::string& algo_spec,
+                           const std::string& provider = "")
+   {
+   std::unique_ptr<PBKDF> p(PBKDF::create(algo_spec, provider));
+   if(p)
```
Comment (webmaster128) [inline src/lib/pbkdf/pbkdf.h]: Can we say that omitting braces is only allowed if the single statement is placed in the same line? [Style Conventions](https://github.com/randombit/botan/blob/master/doc/hacking.rst#style-conventions)

**Comment by webmaster128** [comment]
Great to see unique pointer factories, great to see pbkdf being remove from the base module and - of course – great to see static builds work!

I just started to play around with this by merging it into the cmake branch. With little adjustments, I was able to run the Botan tests in a static cmake build.

File: src/lib/block/block_cipher.cpp [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/block_cipher.h>
+#include <botan/cpuid.h>
+#include <botan/internal/algo_registry.h>
+
+#if defined(BOTAN_HAS_SIMD_32)
+#include <botan/internal/simd_32.h>
```
Comment (webmaster128) [inline src/lib/block/block_cipher.cpp]: Indent?

File: src/lib/block/block_cipher.h [hunk 3]
```diff
@@ -22,6 +22,19 @@ class BOTAN_DLL BlockCipher : public SymmetricAlgorithm
       typedef SCAN_Name Spec;
 
       /**
+      * Create an instance based on a name
+      * Will return a null pointer if the algo/provider combination cannot
+      * be found. If providers is empty then best available is chosen.
```
Comment (webmaster128) [inline src/lib/block/block_cipher.h]: providers -> provider

File: src/lib/hash/hash.h [hunk 4]
```diff
@@ -20,11 +20,30 @@ namespace Botan {
 class BOTAN_DLL HashFunction : public Buffered_Computation
    {
    public:
+      typedef SCAN_Name Spec;
```
Comment (webmaster128) [inline src/lib/hash/hash.h]: Without grepping in the source code, I had no clue that this is supposed to do. What about `using Spec = Botan::SCAN_Name;`?

File: src/lib/modes/aead/aead.cpp [hunk 5]
```diff
@@ -6,9 +6,61 @@
 
 #include <botan/internal/mode_utils.h>
 #include <botan/aead.h>
+#include <botan/lookup.h>
+
+#if defined(BOTAN_HAS_AEAD_CCM)
+#include <botan/ccm.h>
```
Comment (webmaster128) [inline src/lib/modes/aead/aead.cpp]: Indent?

**Comment by webmaster128** [comment]
I think we might need some clarity in factory function consistency. There are

1. HashFunction::create (nothrow, nullable, unique_ptr)
2. make_hash_function (throw, not,-nullable, unique_ptr)
3. get_hash_function (nothrow, nullable, raw pointer)

and

1. BlockCipher::create (nothrow, nullable, unique_ptr)
2. make_block_cipher (throw, not-null, unique_ptr)
3. get_block_cipher (nothrow, nullable, raw pointer)

but the make_* is missing for some cases, e.g. make_kdf, make_pbkdf don't exist:

1. BlockCipher::create (nothrow, nullalble, unique_ptr)
2. get_kdf (**throw**, nullable, raw pointer)

I personally prefer the object oriented style X::create. Checking for nullpointer or exception is no real difference for me. Is there a good reason for either choice? I also see no value in the raw pointers as those can be obtained easily from the unique_ptr.

So I'd opt for checking what we want and deprecate or delete the rest.

File: src/lib/pbkdf/pbkdf.h (hunk 1 above)
Comment (randombit) [inline src/lib/pbkdf/pbkdf.h]: I'd be in favor of adopting this as a style rule.

**Comment by randombit** [comment]
Consistency in the factory functions is non-existent, but the lookup.h signatures (with these inconsistent throw vs null returns) have been in place for some time which is why I kept them here. I'm much happier with the consistency of T::create and think the old functions should probably be deprecated and eventually removed. That is also the reason for not adding make_kdf, etc here.

File: src/lib/pbkdf/pbkdf.h (hunk 1 above)
Comment (webmaster128) [inline src/lib/pbkdf/pbkdf.h]: In my opintion it is too easy to change a code like that

```
if (p)
    p->doOneThink();
```

to

```
if (p)
    p->doOneThink();
    p->doAnotherThink();
```

without any notice by the compiler or anything. That risk can be reduced dramatically by requiring

```
if (p)
   {
   p->doOneThink();
   }
```

or

```
if (p) p->doOneThink();
```

Especially the later versions fits great for null pointer checks like in that case.

Comment (mouse07410) [inline src/lib/pbkdf/pbkdf.h]: +1

________________________________

**Comment by webmaster128** [comment]
Really good job, thanks @randombit!
